pcap.pyx

```diff
@@ -227,92 +237,85 @@ cdef class pcap:
         if pcap_compile(self.__pcap, &fcode, self.__filter, optimize, 0) < 0:
             raise OSError, pcap_geterr(self.__pcap)
         if pcap_setfilter(self.__pcap, &fcode) < 0:
             raise OSError, pcap_geterr(self.__pcap)
         pcap_freecode(&fcode)
-		
+
     def setdirection(self, direction):
         """Set capture direction."""
-        ret = pcap_ex_setdirection(self.__pcap, direction)
-        if ret == 0:
-            return True
-        return False
+        return pcap_ex_setdirection(self.__pcap, direction) == 0
 
     def setnonblock(self, nonblock=True):
         """Set non-blocking capture mode."""
         pcap_ex_setnonblock(self.__pcap, nonblock, self.__ebuf)
-    
+
     def getnonblock(self):
         """Return non-blocking capture mode as boolean."""
         ret = pcap_ex_getnonblock(self.__pcap, self.__ebuf)
         if ret < 0:
             raise OSError, self.__ebuf
-        elif ret:
-            return True
-        return False
-    
+        return ret != 0
+
     def datalink(self):
         """Return datalink type (DLT_* values)."""
         return pcap_datalink(self.__pcap)
-    
+
     def __add_pkts(self, ts, pkt, pkts):
         pkts.append((ts, pkt))
-    
+
     def readpkts(self):
         """Return a list of (timestamp, packet) tuples received in one buffer."""
         pkts = []
         self.dispatch(-1, self.__add_pkts, pkts)
         return pkts
-    
+
     def dispatch(self, cnt, callback, *args):
         """Collect and process packets with a user callback,
         return the number of packets processed, or 0 for a savefile.
-        
+
         Arguments:
-        
+
         cnt      -- number of packets to process;
                     or 0 to process all packets until an error occurs,
                     EOF is reached, or the read times out;
                     or -1 to process all packets received in one buffer
         callback -- function with (timestamp, pkt, *args) prototype
         *args    -- optional arguments passed to callback on execution
         """
-        cdef pcap_handler_ctx ctx
+        cdef pcap_handler_ctx ctx = pcap_handler_ctx()
         cdef int n
 
         ctx.callback = <void *>callback
         ctx.args = <void *>args
-        ctx.got_exc = 0
-        n = pcap_dispatch(self.__pcap, cnt, __pcap_handler,
-                          <unsigned char *>&ctx)
-        if ctx.got_exc:
-            exc = sys.exc_info()
-            raise exc[0], exc[1], exc[2]
+        n = pcap_dispatch(self.__pcap, cnt, __pcap_handler, <u_char *><void*>ctx)
+        if ctx.exc:
+            raise ctx.exc[0], ctx.exc[1], ctx.exc[2]
         return n
 
     def loop(self, cnt, callback, *args):
         """Processing packets with a user callback during a loop.
         The loop can be exited when cnt value is reached
         or with an exception, including KeyboardInterrupt.
-        
+
         Arguments:
 
         cnt      -- number of packets to process;
```

Comment:
TODO: update docstring to mention that actually only `cnt=0` (and not also `-1`) will process all packets until an error occurs or EOF is reached.